src/lib.rs

```diff
@@ -118,6 +122,47 @@ impl<'a, V: VecScopedPrivate> VecScopedPrivate for Push<'a, V> {
 
 impl<'a, T, V: VecScopedPrivate<Element = T>> VecScoped<T> for Push<'a, V> {}
 
+pub struct Update<'a, V: VecScopedPrivate>(&'a mut V, usize, Option<V::Element>);
+
+impl<'a, V: VecScopedPrivate> Update<'a, V> {
+    pub fn new(vec_scoped: &'a mut V, value: V::Element, idx: usize) -> Self {
+        let inner = vec_scoped.vec_mut();
+        // i think this is efficient?
+        let old = inner.swap_remove(idx);
+        let end = inner.len();
+        inner.push(value);
+        inner.swap(end, idx);
+        Self(vec_scoped, idx, Some(old))
+    }
+}
+
+impl<'a, T, V: std::ops::Deref<Target = [T]> + VecScopedPrivate> std::ops::Deref for Update<'a, V> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        std::ops::Deref::deref(self.0)
+    }
+}
+
+impl<'a, V: VecScopedPrivate> Drop for Update<'a, V> {
+    fn drop(&mut self) {
+        let idx = self.1;
+        if let Some(old) = self.2.take() {
```

Comment:
This is interesting because `self.2.take()` should always return `Some`. So, does that mean that `self.2` can actually be a `T` or a `Cell<T>` rather than an `Option`?

Reply:
i couldn't figure out how to do just `T`, because i think i need to take ownership here from the `&mut self` in this function, as it needs to go back into the vec. that's what `Option::take` lets me do, which i guess works to prevent me from taking ownership twice in the case of a bug or something (possibly should panic in an `else` clause?)

it looks like cell lets me do something similar, but i'd need to have a replacement for the `T` in the cell.

it's also possible i'm missing something.

Reply:
ooooooooh wait. you're right, i _can_ get away without an `Option` using `std::mem::swap`, like this: 019378148723997962a5ee01e350456048d3d258